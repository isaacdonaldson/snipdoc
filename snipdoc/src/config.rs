//! This module provides configuration management for the `snipdoc`. It
//! includes functionality to load and manage configurations from a default YAML
//! file.
use std::path::{Path, PathBuf};

use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::errors::ConfigResult;

pub const DEFAULT_CONFIG_NAME: &str = "snipdoc-config.yml";

#[derive(Default, Clone, Debug, Serialize, Deserialize)]
pub struct Config {
    pub walk: WalkConfig,
    pub inject: Option<InjectConfig>,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct InjectConfig {}

#[derive(Default, Clone, Debug, Serialize, Deserialize)]
#[allow(clippy::module_name_repetitions)]
pub struct WalkConfig {
    /// Patterns to include files.
    #[serde(with = "serde_regex", default)]
    pub includes: Vec<Regex>,
    /// Patterns to exclude files.
    #[serde(with = "serde_regex", default)]
    pub excludes: Vec<Regex>,
    /// Patterns to exclude files.
    #[serde(default)]
    pub excludes_file_path: Vec<PathBuf>,
}

impl Config {
    /// Attempts to load the configuration from a default file under the given
    /// path.
    ///
    /// This function first checks if the default configuration file exists at
    /// the given path. If it exists, it tries to load the configuration
    /// from this file. If the file does not exist or contains invalid
    /// content, it falls back to loading the default configuration.
    ///
    /// # Returns
    ///
    /// A [`Config`] instance loaded from the file if it exists and is valid, or
    /// a default `Config` instance otherwise.
    pub fn try_from_default_file(path: &Path) -> Self {
        let maybe_config_exists = path.join(DEFAULT_CONFIG_NAME);

        if maybe_config_exists.exists() {
            if let Ok(config) = Self::from_file(maybe_config_exists.as_path()) {
                tracing::debug!(
                    config = %maybe_config_exists.display(),
                    "configuration file loaded"
                );
                return config;
            }
            tracing::error!(
                config = %maybe_config_exists.display(),
                "invalid config file content"
            );
        } else {
            tracing::debug!(
                config = %maybe_config_exists.display(),
                "default config file not found"
            );
        }
        Self::default()
    }

    /// Loads the configuration from the specified file.
    ///
    /// This function reads the file at the given path and attempts to
    /// deserialize its content into a [`Config`] instance.
    ///
    /// # Errors
    ///
    /// Returns an error if the file cannot be read or if the content is
    /// invalid.
    pub fn from_file(path: &Path) -> ConfigResult<'_, Self> {
        Ok(serde_yaml::from_reader(std::fs::File::open(path)?)?)
    }
}
