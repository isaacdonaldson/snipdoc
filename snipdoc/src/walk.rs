//! This module provides functionality for traversing directories and collecting
//! files, while respecting include and exclude patterns.
//!
//! It utilizes the `ignore` crate for directory traversal and `serde` for
//! serialization and deserialization of configuration patterns.
use std::{
    io,
    path::{Path, PathBuf},
    sync::mpsc,
};

use ignore::WalkBuilder;

use crate::config::WalkConfig;

pub const DEFAULT_CONFIG_NAME: &str = "snipdoc-config.yml";

/// Represents a directory walker with include and exclude configurations.
#[derive(Debug)]
pub struct Walk {
    /// The base folder from which files are collected.
    pub folder: PathBuf,
    config: WalkConfig,
}

impl Walk {
    /// Creates a new instance of [`Walk`] with the specified base folder.
    ///
    /// # Errors
    ///
    /// Returns an error if the provided folder path is invalid.
    pub fn new(folder: &Path) -> io::Result<Self> {
        Ok(Self {
            folder: dunce::canonicalize(folder)?,
            config: WalkConfig::default(),
        })
    }

    /// Creates a [`Walk`] instance from the given configuration.
    ///
    /// # Errors
    ///
    /// Returns an error if the provided folder path is invalid.
    pub fn from_config(folder: &Path, config: &WalkConfig) -> io::Result<Self> {
        Ok(Self {
            folder: dunce::canonicalize(folder)?,
            config: config.clone(),
        })
    }

    /// Checks if a file should be excluded based on configured exclude
    /// patterns.
    fn should_exclude(&self, path: &Path) -> bool {
        let path = path
            .strip_prefix(&self.folder)
            .unwrap()
            .display()
            .to_string();

        for exclude in &self.config.excludes {
            if exclude.is_match(&path) {
                tracing::trace!("file excluded from configurations");
                return true;
            }
        }
        false
    }

    /// Checks if a file should be included based on configured include
    /// patterns.
    fn should_include(&self, path: &Path) -> bool {
        let path = path
            .strip_prefix(&self.folder)
            .unwrap()
            .display()
            .to_string();

        if self.config.includes.is_empty() {
            return true;
        }

        for include in &self.config.includes {
            if include.is_match(&path) {
                tracing::trace!("file excluded from configurations");
                return true;
            }
        }
        tracing::debug!("file should not be included");
        false
    }

    fn exclude_file_path(&self, path: &Path) -> bool {
        for exclude_file in &self.config.excludes_file_path {
            let absolute_path =
                dunce::canonicalize(exclude_file).unwrap_or_else(|_| exclude_file.clone());
            if absolute_path == path {
                tracing::debug!("file should not be included");
                return true;
            }
        }

        false
    }

    /// Collects files in the specified folder, respecting exclude and include
    /// patterns.
    #[must_use]
    pub fn get_files(&self) -> Vec<PathBuf> {
        let (tx, rx) = mpsc::channel();
        WalkBuilder::new(&self.folder)
            .build_parallel()
            .run(move || {
                let tx = tx.clone();
                Box::new(move |result| {
                    result.map_or_else(
                        |err| {
                            tracing::error!(err = %err,"dir entry error ");
                        },
                        |entry| {
                            if entry.path().is_file() {
                                let path = entry.path().to_owned();
                                if !self.exclude_file_path(path.as_path()) && !self.should_exclude(path.as_path()) && self.should_include(path.as_path()){
                                    if let Err(err) = tx.send(path.clone()) {
                                        tracing::error!(err = %err,path = %path.display(),"error sending path to tx ");
                                    }
                                }
                            }
                        },
                    );
                    ignore::WalkState::Continue
                })
            });

        rx.into_iter().collect::<Vec<_>>()
    }
}
