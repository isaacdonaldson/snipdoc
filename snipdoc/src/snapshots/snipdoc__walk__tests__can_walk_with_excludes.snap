---
source: snipdoc/src/walk.rs
expression: file_paths
---
[
    "file.txt",
    "folder/file.txt",
    "folder/folder/2.txt",
    "README.md",
]
